cupy/linalg/einsum.py

```diff
@@ -305,45 +438,213 @@ def einsum(*operands):
+            dim = 1
+            upper = ellipsis_pos + ioperand.ndim - len(subscript) + 1
+            for j in six.moves.range(ellipsis_pos, upper):
+                dim *= ioperand.shape[j]
+            broadcasted_dims.append(dim)
+    if broadcasted_dims:
+        dim_dict['@'] = max(broadcasted_dims)
+
+    # Compute size of each input array plus the output array
+    size_list = []
+    for subscript in input_list + [output_subscript]:
+        size_list.append(_compute_size_by_dict(subscript, dim_dict,
+                                               broadcasted_dims))
+    max_size = max(size_list)
+
+    if memory_limit is None:
+        memory_arg = max_size
+    else:
+        memory_arg = memory_limit
+
+    # Compute the path
+    if not path_type or (len(input_list) in [1, 2]) or (indices == output_set):
+        # Nothing to be optimized, leave it to einsum
+        path = [tuple(six.moves.range(len(input_list)))]
+    elif path_type == 'greedy':
+        # Maximum memory should be at most out_size for this algorithm
+        memory_arg = min(memory_arg, max_size)
+        path = _greedy_path(input_sets, output_set, dim_dict, broadcasted_dims,
+                            memory_arg)
+    elif path_type[0] == 'einsum_path':
+        path = path_type[1:]
+    else:
+        raise KeyError('Path name %s not found', path_type)
 
-    result, subscript = calc_summed_view(result, subscript, output_subscript)
-    result = calc_transposed_view(result, subscript, output_subscript)
-    return result
+    cost_list, scale_list, size_list, contraction_list = [], [], [], []
+
+    # Build contraction tuple (positions, gemm, einsum_str, remaining)
+    for cnum, contract_inds in enumerate(path):
+        # Make sure we remove inds from right to left
+        contract_inds = tuple(sorted(list(contract_inds), reverse=True))
+
+        contract = _find_contraction(contract_inds, input_sets, output_set)
+        out_inds, input_sets, idx_removed, idx_contract = contract
+
+        cost = _compute_size_by_dict(idx_contract, dim_dict, broadcasted_dims)
+        if idx_removed:
+            cost *= 2
+        cost_list.append(cost)
+        scale_list.append(len(idx_contract))
+        size_list.append(_compute_size_by_dict(out_inds, dim_dict,
+                                               broadcasted_dims))
+
+        tmp_inputs = []
+        for x in contract_inds:
+            tmp_inputs.append(input_list.pop(x))
+
+        # Last contraction
+        if cnum == len(path) - 1:
+            idx_result = output_subscript
+        else:
+            sort_result = [(dim_dict[ind], ind) for ind in out_inds]
+            idx_result = ''.join([x[1] for x in sorted(sort_result)])
+
+        input_list.append(idx_result)
+        einsum_str = ','.join(tmp_inputs) + '->' + idx_result
+
+        contraction = (contract_inds, idx_removed, einsum_str, input_list[:])
+        contraction_list.append(contraction)
+
+    return operands, contraction_list
```

Comment:
It looks that the function always behaves like `einsum_call = True`, right? So it is not compatible with the original one. I think you just want to optimize `einsum`, and it is not for users. It is better to add a comment about it in the docstring.